### PR TITLE
[Parser] Initial implementation of Statements

### PR DESCRIPTION
**Pull request by Thirumalai-Shaktivel**
Related: #298

### CODE REVIEW

**Comment by Thirumalai-Shaktivel** [comment]
Used this docs for reference: https://docs.python.org/3/reference/simple_stmts.html#simple-statements

**Comment by certik** [comment]
I think that looks good so far.

We should probably add tests into `run_tests.py`, such as `ast_new`, which would test the new parser. That way we can test the parser step by step until it is done.

**Comment by Thirumalai-Shaktivel** [comment]
Yup, This is perfect idea. I will do it

**Comment by Thirumalai-Shaktivel** [comment]
Before that I should resolve the error which says:
```console
$ lpython --show-ast --new-parser examples/expr2.py
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  Binary file "/home/thirumalai/Open_Source/lpython/src/bin/lpython", in _start()
  Binary file "/lib/x86_64-linux-gnu/libc.so.6", in __libc_start_main()
  File "/home/thirumalai/Open_Source/lpython/src/bin/lpython.cpp", line 752, in ??
    return emit_ast(arg_file, runtime_library_dir, compiler_options);
  File "/home/thirumalai/Open_Source/lpython/src/bin/lpython.cpp", line 122, in ??
    al, runtime_library_dir, infile, diagnostics, compiler_options.new_parser);
  File "/home/thirumalai/Open_Source/lpython/src/lpython/parser/parser.cpp", line 95, in LFortran::parse_python_file(Allocator&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LFortran::diag::Diagnostics&, bool)
    Result<LPython::AST::Module_t*> res = parse(al, input, diagnostics);
  File "/home/thirumalai/Open_Source/lpython/src/lpython/parser/parser.cpp", line 20, in LFortran::parse(Allocator&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LFortran::diag::Diagnostics&)
    p.parse(s);
  File "/home/thirumalai/Open_Source/lpython/src/lpython/parser/parser.cpp", line 52, in LFortran::Parser::parse(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
    if (yyparse(*this) == 0) {
  File "/home/thirumalai/Open_Source/lpython/src/lpython/parser.yy", line 229, in yyparse(LFortran::Parser&)
AssertFailed: is_a<T>(*f)
```

File: src/lpython/parser/semantics.h [hunk 1]
```diff
@@ -19,19 +19,70 @@
...
+}
+
+#define VEC_CAST(x, type) vec_cast<type##_t, astType::type>(x)
+#define STMTS(x) VEC_CAST(x, stmt)
+#define EXPRS(x) VEC_CAST(x, expr)
 
 #define EXPR(x) (down_cast<expr_t>(x))
 #define EXPR2STMT(x) ((stmt_t*)make_Expr_t(p.m_a, x->base.loc, x))
 
 #define RESULT(x) p.result.push_back(p.m_a, EXPR2STMT(EXPR(x)))
```
Comment (Thirumalai-Shaktivel) [inline src/lpython/parser/semantics.h:52]: I think the error was caused by this, but I'm not sure. I need to debug and fix it.

Comment (Thirumalai-Shaktivel) [inline src/lpython/parser/semantics.h:52]: The error is fixed!!
```suggestion
#define STMT(x) (down_cast<stmt_t>(x))
#define RESULT(x) p.result.push_back(p.m_a, STMT(EXPR(x)))
```

**Comment by certik** [comment]
Great job. Please ping me once this is ready for review.

**Comment by Thirumalai-Shaktivel** [comment]
The only left out part is the multiline statements.

**Comment by Thirumalai-Shaktivel** [comment]
This PR is ready to go in!

File: src/lpython/parser/tokenizer.re [hunk 2]
```diff
@@ -67,27 +75,23 @@ void lex_int(Allocator &al, const unsigned char *s,
...
-            s++;
-        } else {
-            s = s + 2;
-        }
+        s = s + 2;
         prefix.p = (char*) "Oct";
         prefix.n = 3;
         Str num;
-        num.p = (char*)s;
+        num.p = get_value(al, (char*)s, 8);
```
Comment (certik) [inline src/lpython/parser/tokenizer.re:94]: That's fine for now, but it's not very efficient. We should convert directly without an intermediate string.

**Review by certik — APPROVED** [review]
This looks good. Many thanks for this huge improvement to the parser. We will keep iterating on this.

**Comment by Thirumalai-Shaktivel** [comment]
Thanks for reviewing and merging this PR!

**Pull request merged**